File: c_socket/integration_test.c
#include "c_socket_packets.h"
#include "c_socket.h"
#include <stdio.h>
#include <stdlib.h>

int main() {
  
  c_socket_init();
  c_socket_bind_and_listen(5560);
  c_socket_accept();
  void* packet = malloc(65536);
  int type;

  // receive hello
  type = c_socket_receive_packet(packet);
  printf("greeting: %s\n", ((struct HelloPacket*) packet)->greeting);
  printf("expecting: hello\n");
  printf("\n");

  // receive hello reply
  type = c_socket_receive_packet(packet);
  printf("greeting: %s\n", ((struct HelloReplyPacket*) packet)->greeting);
  printf("expecting: helloreply\n");
  printf("\n");

  // receive map
  type = c_socket_receive_packet(packet);
  printf("map[0]: %d\n", ((struct MapPacket*) packet)->map[0]);
  printf("expecting: 15\n");
  printf("\n");

  // receive unit info
  type = c_socket_receive_packet(packet);
  printf("unit_id: %d\n", ((struct UnitInfoPacket*) packet)->unit_id);
  printf("expecting: 13\n");
  printf("\n");

  // receive civ info
  type = c_socket_receive_packet(packet);
  printf("nation_tag: %d\n", ((struct CivInfoPacket*) packet)->nation_tag);
  printf("expecting: 20\n");
  printf("\n");

  // receive city info
  type = c_socket_receive_packet(packet);
  printf("city_name: %s\n", ((struct CityInfoPacket*) packet)->city_name);
  printf("expecting: city\n");
  printf("pop: %d\n", ((struct CityInfoPacket*) packet)->pop);
  printf("expecting: 100\n");
  printf("owned_by: %s\n", ((struct CityInfoPacket*) packet)->owned_by);
  printf("expecting: me\n");
  printf("\n");

  // receive action
  type = c_socket_receive_packet(packet);
  printf("action: %s\n", ((struct ActionPacket*) packet)->action);
  printf("expecting: action\n");
  printf("action_specifiers: %s\n", ((struct ActionPacket*) packet)->action_specifiers);
  printf("expecting: magestically\n");
  printf("\n");

  // receive action reply
  type = c_socket_receive_packet(packet);
  printf("action: %s\n", ((struct ActionReplyPacket*) packet)->action);
  printf("expecting: reply\n");
  printf("\n");

  // receive turn begin
  type = c_socket_receive_packet(packet);
  printf("turn_begin: %d\n", ((struct TurnBeginPacket*) packet)->turn_begin);
  printf("expecting: 18\n");
  printf("\n");

  // receive turn end
  type = c_socket_receive_packet(packet);
  printf("turn_end: %s\n", ((struct TurnEndPacket*) packet)->turn_end);
  printf("expecting: end\n");
  printf("\n");

  // receive completed state transfer
  type = c_socket_receive_packet(packet);
  printf("done: %s\n", ((struct CompletedStateTransferPacket*) packet)->done);
  printf("expecting: yeah\n");
  printf("\n");

  free(packet);

  // send hello
  struct HelloPacket hello = {
    .greeting = "hi"
  };
  c_socket_send_hello_packet(&hello);

  // send hello reply
  struct HelloReplyPacket hello_reply = {
    .greeting = "hi reply"
  };
  c_socket_send_hello_reply_packet(&hello_reply);

  // send map
  struct MapPacket map = {
    .map = { 3 }
  };
  c_socket_send_map_packet(&map);

  // send unit info
  struct UnitInfoPacket unit_info = {
    .unit_id = 5
  };
  c_socket_send_unit_info_packet(&unit_info);

  // send civ info
  struct CivInfoPacket civ_info = {
    .nation_tag = 7
  };
  c_socket_send_civ_info_packet(&civ_info);

  // send city info
  struct CityInfoPacket city_info = {
    .city_name = "city city",
    .pop = 123,
    .owned_by = "me me"
  };
  c_socket_send_city_info_packet(&city_info);

  // send action
  struct ActionPacket action = {
    .action = "doing the thing",
    .action_specifiers = "quickly"
  };
  c_socket_send_action_packet(&action);

  // send action reply
  struct ActionReplyPacket action_reply = {
    .action = "im tired"
  };
  c_socket_send_action_reply_packet(&action_reply);

  // send turn begin
  struct TurnBeginPacket turn_begin = {
    .turn_begin = 13
  };
  c_socket_send_turn_begin_packet(&turn_begin);

  // send turn end
  struct TurnEndPacket turn_end = {
    .turn_end = "turn ended wake up"
  };
  c_socket_send_turn_end_packet(&turn_end);

  // send completed state transfer
  struct CompletedStateTransferPacket completed_state_transfer = {
    .done = "finally"
  };
  c_socket_send_completed_state_transfer_packet(&completed_state_transfer);



  c_socket_close();

  return 0;
}

File: PythonClient/integration_test.py
import socket
from socket_listener import SocketClient
from packets import *


if __name__ == '__main__':
    ip, port = 'localhost', 5560
    listener = SocketClient(ip, port)

    # send hello
    hello = HelloPacket()
    hello.set_content('greeting', 'hello')
    listener.send_packet(hello)

    # send hello reply
    hello_reply = HelloReplyPacket()
    hello_reply.set_content('greeting', 'helloreply')
    listener.send_packet(hello_reply)

    # send map
    mapp = MapPacket()
    mapp.set_content('map', [15] * 4096)
    listener.send_packet(mapp)

    # send unit info
    unit_info = UnitInfoPacket()
    unit_info.set_content('unit_id', 13)
    unit_info.set_content('owner', 'owner')
    unit_info.set_content('nationality', 'nation')
    unit_info.set_content('coordx', 15)
    unit_info.set_content('coordy', 51)
    unit_info.set_content('upkeep', 16)
    listener.send_packet(unit_info)

    # send civ info
    civ_info = CivInfoPacket()
    listener.send_packet(civ_info)

    # send city info
    city_info = CityInfoPacket()
    city_info.set_content('id', 100)
    city_info.set_content('coordx', 101)
    city_info.set_content('coordy', 10)
    city_info.set_content('owner', 102)
    city_info.set_content('size', 103)
    city_info.set_content('radius', 104)
    city_info.set_content('food_stock', 105)
    city_info.set_content('shield_stock', 106)
    city_info.set_content('production_kind', 107)
    city_info.set_content('production_value', 108)
    city_info.set_content('improvements', 'improve')
    listener.send_packet(city_info)

    # send action
    action = ActionPacket()
    action.set_content('action', 'action')
    action.set_content('ACTION_ID', 5)
    action.set_content('actor_id', 6)
    action.set_content('target_id', 7)
    listener.send_packet(action)

    # send action reply
    action_reply = ActionReplyPacket()
    action_reply.set_content('action', 'reply')
    listener.send_packet(action_reply)

    # send turn begin
    turn_begin = TurnBeginPacket()
    turn_begin.set_content('turn_begin', 18)
    listener.send_packet(turn_begin)

    # send turn end
    turn_end = TurnEndPacket()
    turn_end.set_content('turn_end', 'end')
    listener.send_packet(turn_end)

    # send completed state transfer
    completed_state_transfer = CompletedStateTransferPacket()
    completed_state_transfer.set_content('done', 'yeah')
    listener.send_packet(completed_state_transfer)

    # send completed state transfer
    research_info = ResearchInfoPacket()
    research_info.set_content('id', 17)
    research_info.set_content('techs_researched', 18)
    research_info.set_content('researching', 'tech')
    research_info.set_content('researching_cost', 19)
    research_info.set_content('bulbs_researched', 20)
    listener.send_packet(research_info)



    # receive hello
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive hello reply
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive map
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive unit info
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive civ info
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive city info
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive action
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive action reply
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive turn begin
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive turn end
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive completed state transfer
    packet = listener.receive_packet()
    print(packet.content)
    print()

    # receive research info
    packet = listener.receive_packet()
    print(packet.content)
    print()

    listener.close()
